mlfinlab/portfolio_optimization/nco.py

```diff
@@ -0,0 +1,362 @@
+            w_nco.loc[simulation] = self.opt_port_nco(cov_simulation, mu_simulation,
+                                                      int(cov_simulation.shape[0] / 2)).flatten()
+
+        return w_cvo, w_nco
+
+    def estim_errors_mcos(self, w_cvo, w_nco, mu_vec, cov, min_var_portf):
+        """
+        Computes the true optimal allocation w, and compares that result with the estimated ones by MCOS.
+
+        The result is the mean standard deviation between the true weights and the ones obtained from the simulation
+        for each algorithm - CVO and NCO.
+
+        :param w_cvo: (pd.dataframe) DataFrame with weights from the CVO algorithm.
+        :param w_nco: (pd.dataframe) DataFrame with weights from the NCO algorithm.
+        :param mu_vec: (np.array) The original vector of expected outcomes.
+        :param cov: (np.array)The original covariance matrix of outcomes.
+        :param min_var_portf: (bool) When True, the minimum variance solution was computed. Otherwise, the
+                                     maximum Sharpe ratio solution was computed.
+        :return: (float, float) Mean standard deviation of weights for CVO and NCO algorithms.
+        """
+
+        # Calculating the true optimal weights allocation
+        w_true = self.opt_port(cov, None if min_var_portf else mu_vec)
+        w_true = np.repeat(w_true.T, w_cvo.shape[0], axis=0)
+
+        # Mean standard deviation between the weights from CVO algorithm and the true weights
+        err_cvo = (w_cvo - w_true).std(axis=0).mean()
+
+        # Mean standard deviation between the weights from NCO algorithm and the true weights
+        err_nco = (w_nco - w_true).std(axis=0).mean()
+
+        return err_cvo, err_nco
+
+    @staticmethod
+    def _form_block_matrix(num_blocks, block_size, block_corr):
+        """
+        Creates a correlation matrix in a block form with given parameters.
+
+        :param num_blocks: (int) Number of blocks in matrix
+        :param block_size: (int) Size of a single block
+        :param block_corr: (float) Correlation of elements in a block
+        :return: (np.array) Resulting correlation matrix
+        """
+
+        # Creating a single block with all elements as block_corr
+        block = np.ones((block_size, block_size)) * block_corr
+
+        # Setting the main diagonal to ones
+        block[range(block_size), range(block_size)] = 1
+
+        # Create a block diagonal matrix with a number of equal blocks
+        res_matrix = block_diag(*([block] * num_blocks))
+
+        return res_matrix
+
+    def form_true_matrix(self, num_blocks, block_size, block_corr, std=None):
+        """
+        Creates a random vector of means and a random covariance matrix.
+
+        Due to the block structure of a matrix, it is a good sample data to use in the NCO and MCOS algorithms.
+
+        The number of securities in a portfolio, number of blocks and correlations
```

Comment:
"The number of securities in a portfolio" -> "The number of assets in a portfolio"

Reply:
Done